File: leetcode/python/LEET938.py
from typing import *
import collections

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def rangeSumBST(root: TreeNode, low: int, high: int) -> int:
    def dfs(node: TreeNode):
        if not node:
            return 0
        
        if node.val > high:
            return dfs(node.left)
        
        if node.val < low:
            return dfs(node.right)
        
        return node.val + dfs(node.right) + dfs(node.left)

def solve():
    root = build_tree([15,9,21,7,13,19,23,5,None,11,None,17])
    res = rangeSumBST(root,low = 19, high = 21)
    print(res)

def build_tree(array: List[int]) -> TreeNode:
    if not array:
        return None

    root = TreeNode(array[0])
    q = collections.deque([root])
    index = 1

    while index < len(array):
        node = q.popleft()
        if array[index] is not None:
            left = TreeNode(array[index])
            node.left = left
            q.append(left)
        index += 1
        
        if index >= len(array): break

        if array[index] is not None:
            right = TreeNode(array[index])
            node.right = right
            q.append(right)
        index += 1

    return root

def print_tree(root):
    q = collections.deque([root])

    while q:
        node = q.popleft()
        if node:
            print(node.val, end=" ")
        else:
            print(None, end=" ")
            continue
        
        q.append(node.left)
        q.append(node.right)
    print()
solve()

File: leetcode/java/LEET938.java
import java.util.*;

public class LEET938 {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public static void main(String[] args) {
        TreeNode root = buildTree(new Integer[] {10,5,15,3,7,null,18});
        System.out.println(rangeSumBST(root, 7, 15));
    }
    public static int rangeSumBST(TreeNode root, int low, int high) {
        return dfs(root, low, high);
    }
    public static int dfs(TreeNode node, int low, int high){
        if(node == null)
            return 0;

        if(node.val > high)
            return dfs(node.left, low, high);
        if(node.val < low)
            return dfs(node.right, low, high);

        return node.val + dfs(node.left, low, high) + dfs(node.right, low, high);
    }
    public static void printTree(TreeNode root){
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            TreeNode node = q.poll();
            if(node == null){
                System.out.print("NULL ");
                continue;
            }
            System.out.print(node.val +" ");
            
            q.add(node.left);
            q.add(node.right);
        }
        System.out.println();
    }
    public static TreeNode buildTree(Integer[] nums){
        if(nums.length == 0){
            return null;
        }
        Queue<TreeNode> q = new LinkedList<>();
        TreeNode root = new TreeNode(nums[0]);
        int index = 1;
        q.add(root);
        while(index < nums.length){
            TreeNode node = q.poll();
            if(nums[index] != null){
                node.left = new TreeNode(nums[index]);
                q.add(node.left);
            }
            index++;

            if(index >= nums.length)
                break;
            
            if(nums[index] != null){
                node.right = new TreeNode(nums[index]);
                q.add(node.right);
            }
            index++;
        }
        return root;
    }
}
